rcl/src/rcl/wait.c

```diff
@@ -614,6 +614,154 @@ rcl_wait(rcl_wait_set_t * wait_set, int64_t timeout)
+        rmw_guard_condition_t * rmw_handle = rcl_guard_condition_get_rmw_handle(guard_condition);
+        collated_wait_set.impl->rmw_guard_conditions.guard_conditions[index] = rmw_handle->data;
+      }
+    }
+    for (j = 0u; j < wait_set->size_of_subscriptions; ++j) {
+      const size_t current_index = collated_wait_set.impl->subscription_index++;
+      collated_wait_set.subscriptions[current_index] = wait_set->subscriptions[j];
+      rmw_subscription_t * rmw_handle = rcl_subscription_get_rmw_handle(wait_set->subscriptions[j]);
+      collated_wait_set.impl->rmw_subscriptions.subscribers[current_index] = rmw_handle->data;
+    }
+    for (j = 0u; j < wait_set->size_of_guard_conditions; ++j) {
+      const size_t current_index = collated_wait_set.impl->guard_condition_index++;
+      collated_wait_set.guard_conditions[current_index] = wait_set->guard_conditions[j];
+      rmw_guard_condition_t * rmw_handle = rcl_guard_condition_get_rmw_handle(
+        wait_set->guard_conditions[j]);
+      collated_wait_set.impl->rmw_guard_conditions.guard_conditions[current_index] =
+        rmw_handle->data;
+    }
+    for (j = 0u; j < wait_set->size_of_clients; ++j) {
+      const size_t current_index = collated_wait_set.impl->client_index++;
+      collated_wait_set.clients[current_index] = wait_set->clients[j];
+      rmw_client_t * rmw_handle = rcl_client_get_rmw_handle(wait_set->clients[j]);
+      collated_wait_set.impl->rmw_clients.clients[current_index] = rmw_handle->data;
+    }
+    for (j = 0u; j < wait_set->size_of_services; ++j) {
+      const size_t current_index = collated_wait_set.impl->service_index++;
+      collated_wait_set.services[current_index] = wait_set->services[j];
+      rmw_service_t * rmw_handle = rcl_service_get_rmw_handle(wait_set->services[j]);
+      collated_wait_set.impl->rmw_services.services[current_index] = rmw_handle->data;
+    }
+  }
+  collated_wait_set.impl->rmw_subscriptions.subscriber_count = number_of_subscriptions;
+  collated_wait_set.impl->rmw_guard_conditions.guard_condition_count = number_of_guard_conditions;
+  collated_wait_set.impl->rmw_clients.client_count = number_of_clients;
+  collated_wait_set.impl->rmw_services.service_count = number_of_services;
+
+  // Wait on the collated wait set
+  rcl_ret_t wait_ret = rcl_wait(&collated_wait_set, timeout);
+  if (RCL_RET_OK != wait_ret && RCL_RET_TIMEOUT != wait_ret) {
+    // Finalize collated wait set
+    ret = rcl_wait_set_fini(&collated_wait_set);
+    (void)ret;
+    return RCL_RET_ERROR;  // error already set
+  }
+
+  // Update individual wait sets
+  size_t collated_subscription_index = 0u;
+  size_t collated_guard_condition_index = 0u;
+  size_t collated_timer_index = 0u;
+  size_t collated_client_index = 0u;
+  size_t collated_service_index = 0u;
+  for (i = 0u; i < num_wait_sets; ++i) {
+    wait_set = &wait_sets[i];
+    for (j = 0u; j < wait_set->size_of_timers; ++j) {
```

Comment:
@jacobperron nit: fine to use `for` loops, but we could use `memcpy` too (that depending on the compiler and target architecture may be more performant).

Reply:
Sure, I can do that. Do you think there is any issues with readability (and correctness)?

```c
    memcpy(
      wait_set->timers,
      collated_wait_set.timers + collated_timer_index,
      sizeof(rcl_timer_t *) * wait_set->size_of_timers);
```

Reply:
Ugh, that doesn't look nearly as clean as your current code. Nevermind, let's trust loop unrolling will work :).

Reply:
:+1: